rust/lance-index/src/vector/graph.rs

```diff
@@ -142,10 +147,82 @@ pub trait Graph {
     }
 
     /// Get the neighbors of a graph node, identifyied by the index.
-    fn neighbors(&self, key: u32) -> Option<Box<dyn Iterator<Item = u32> + '_>>;
+    fn neighbors(&self, key: u32) -> Arc<Vec<u32>>;
+}
+
+/// Array-based visited list (faster than HashSet)
+pub struct Visited<'a> {
+    visited: MutexGuard<'a, Vec<bool>>,
+    recently_visited: Vec<u32>,
+}
+
+impl<'a> Visited<'a> {
+    pub fn insert(&mut self, node_id: u32) {
+        let node_id_usize = node_id as usize;
```

Comment:
Actually very curious about the overall throughput vs CPU cores here. 
Will the insert threads have a lot of contention ?

Reply:
we have a pool of `VisitedGenerator` which is with enough size, see https://github.com/BubbleCal/lance/blob/merge-hnsw/rust/lance-index/src/vector/hnsw/builder.rs#L518

Reply:
We have build() using visited generators unique to each thread, rather than using a pool in this case. I'm also curious what the visited generator contention would be if you've tested it without the new builds though.